numpy/core/arrayprint.py

```diff
@@ -641,53 +647,58 @@ def fillFormat(self, data):
 
         if self.exp_format:
             self.large_exponent = 0 < min_val < 1e-99 or max_val >= 1e100
-            self.max_str_len = 8 + self.precision
-            if self.large_exponent:
-                self.max_str_len += 1
-            if self.sign:
-                format = '%+'
-            else:
-                format = '%'
-            format = format + '%d.%de' % (self.max_str_len, self.precision)
+
+            signpos = self.sign != '-' or _nc.any(non_zero < 0)
+            # for back-compatibility with np 1.13, add extra space if padded
+            signpos = signpos if self.sign != ' ' else 2
+            max_str_len = signpos + 6 + self.precision + self.large_exponent
+
+            signchar = '' if self.sign == '-' else self.sign
+            format = '%' + signchar + '%d.%de' % (max_str_len, self.precision)
         else:
-            format = '%%.%df' % (self.precision,)
-            if len(non_zero):
-                precision = max([_digits(x, self.precision, format)
-                                 for x in non_zero])
+            if len(non_zero) and self.precision > 0:
+                precision = self.precision
+                trim_zero = lambda s: precision - (len(s) - len(s.rstrip('0')))
+                fmt = '%%.%df' % (precision,)
+                precision = max(trim_zero(fmt % x) for x in abs_non_zero)
             else:
                 precision = 0
-            precision = min(self.precision, precision)
-            self.max_str_len = len(str(int(max_val))) + precision + 2
+
+            int_len = len(str(int(max_val)))
+            signpos = self.sign != '-' or (len(str(int(min_val_sgn))) > int_len)
+            max_str_len = signpos + int_len + 1 + precision
+
             if _nc.any(special):
-                self.max_str_len = max(self.max_str_len,
-                                       len(_nan_str),
-                                       len(_inf_str)+1)
-            if self.sign:
-                format = '%#+'
-            else:
-                format = '%#'
-            format = format + '%d.%df' % (self.max_str_len, precision)
+                neginf = self.sign != '-' or _nc.any(data[hasinf] < 0)
+                nanlen = len(_format_options['nanstr'])
+                inflen = len(_format_options['infstr']) + neginf
+                max_str_len = max(max_str_len, nanlen, inflen)
+
+            signchar = '' if self.sign == '-' else self.sign
+            format = '%#' + signchar + '%d.%df' % (max_str_len, precision)
```

Comment:
From https://docs.python.org/3/library/string.html#format-specification-mini-language, it seems the sign character should precede `#`.

Reply:
Not sure if this a python bug, but using '-' in the python format specification causes the number to be left-justified.
```python
>>> '% #8.1f' % 1.0
'     1.0'
>>> '%+#8.1f' % 1.0
'    +1.0'
>>> '%-#8.1f' % 1.0
'1.0     '
```

Reply:
oops that refers to your previsous comment on signchar

Reply:
It's not a bug, it's what C does too.

Reply:
Very odd:
```
In [4]: '%-#8.1f' % 1.
Out[4]: '1.0     '

In [5]: '{:-#8.1f}'.format(1.)
Out[5]: '     1.0'
```
I guess there is a difference between old and new-style formats (I only use the latter, should have checked before commenting).

Reply:
Also, the docs you linked to are for the `.format()` mini-language, not for the printf-style formatting.

Printf-style formatting is documented at https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting ,  which indeed says that '-' left fustifies, and also does not require any order to the conversion flags.

Reply:
Here I was thinking the two are the same... Learn something new every day.

Reply:
Actually, while working on this PR at some point I tried to replace the printf-style formatting by the `.format()` function. I don't remember exactly what was the problem anymore, but I remember I determined that it is actually impossible to format things the way we want using `format()`, and `printf` formatting was more powerful in this case.